crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,310 @@
+    }
+  }
+  return bytes_output;
+}
+
+static int enc_flush(BIO *b, BIO *next, BIO_ENC_CTX *ctx) {
+  GUARD_PTR(b);
+  GUARD_PTR(next);
+  GUARD_PTR(ctx);
+  while (ctx->buf_len > 0 || !ctx->done) {
+    int bytes_written = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (ctx->buf_len > 0 && bytes_written <= 0) {
+      BIO_copy_next_retry(b);
+      return 0;
+    }
+    ctx->buf_off += bytes_written;
+    ctx->buf_len -= bytes_written;
+    if (ctx->buf_len == 0 && !ctx->done) {
+      ctx->done = 1;
+      ctx->buf_off = 0;
+      ctx->ok = EVP_EncryptFinal_ex(ctx->cipher, ctx->buf, &ctx->buf_len);
+    }
+  }
+  return ctx->ok;
+}
+
+static int enc_write(BIO *b, const char *in, int inl) {
+  GUARD_PTR(b);
+  GUARD_PTR(in);
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL || ctx->cipher == NULL || ctx->done || !ctx->ok) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  int bytes_consumed = 0;
+  int remaining = inl;
+  const int max_crypt_size =
+      (int)sizeof(ctx->buf) - EVP_CIPHER_CTX_block_size(ctx->cipher) + 1;
+  while (remaining > 0) {
+    assert(bytes_consumed + remaining == inl);
+    if (ctx->buf_len == 0) {
+      ctx->buf_off = 0;
+      int to_encrypt = remaining < max_crypt_size ? remaining : max_crypt_size;
+      uint8_t *in_pos = ((uint8_t *)in) + bytes_consumed;
+      ctx->ok = EVP_EncryptUpdate(ctx->cipher, ctx->buf, &ctx->buf_len, in_pos,
+                                  to_encrypt);
+      if (!ctx->ok) {
+        break;
+      };
+      bytes_consumed += to_encrypt;
+      remaining -= to_encrypt;
+    }
+    int bytes_written = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (bytes_written <= 0) {
+      BIO_copy_next_retry(b);
+      break;
+    }
+    ctx->buf_off += bytes_written;
+    ctx->buf_len -= bytes_written;
+  }
+  return bytes_consumed;
+}
+
+static long enc_ctrl(BIO *b, int cmd, long num, void *ptr) {
+  GUARD_PTR(b);
+  long ret = 1;
+
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  BIO *next = BIO_next(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  switch (cmd) {
+    case BIO_CTRL_RESET:
+      ctx->done = 0;
+      ctx->ok = 1;
+      ctx->buf_off = 0;
+      ctx->buf_len = 0;
+      OPENSSL_cleanse(ctx->buf, sizeof(ctx->buf));
+      if (!EVP_CipherInit_ex(ctx->cipher, NULL, NULL, NULL, NULL,
+                             EVP_CIPHER_CTX_encrypting(ctx->cipher))) {
+        return 0;
+      }
+      ret = BIO_ctrl(next, cmd, num, ptr);
+      break;
+    case BIO_CTRL_EOF:
+      if (ctx->done) {
+        ret = 1;
+      } else {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_CTRL_WPENDING:
+    case BIO_CTRL_PENDING:
+      // Return number of bytes left to process if we have anything buffered,
+      // else consult underlying BIO.
+      ret = ctx->buf_len;
+      if (ret <= 0) {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
```

Comment:
Should the return value be the sum of the two values? 
```
...
    ret = ctx->buf_len;
    int next_pending = BIO_ctrl(next, cmd, num, ptr);
    // sanity check next_pending
    ret += next_pending;
    break;
```

Reply:
OpenSSL docs excerpted below, not particularly clear on the question 

> BIO_pending(), BIO_ctrl_pending(), BIO_wpending() and BIO_ctrl_wpending() return the amount of pending data. BIO_pending() and BIO_wpending() return negative value or 0 on error.

OpenSSL [has same behavior as us](https://github.com/openssl/openssl/blob/f30d6ba455e06572250e75132045eedde5d1daf0/crypto/evp/bio_enc.c#L327-L329), so i think we should keep it as-is for compatibility.